src/mpl2/src/clusterEngine.cpp

```diff
@@ -1452,16 +1464,59 @@ void ClusteringEngine::breakLargeFlatCluster(Cluster* parent)
                                               vertex_weight,
                                               hyperedge_weights);
 
+  // Check if the partitioning was successful
+  auto checkBalance = [&]() -> bool {
+    float balance_part_0 = 0.0;
+    float balance_part_1 = 0.0;
+    for (int i = 0; i < num_vertices; i++) {
+      if (part[i] == 0) {
+        balance_part_0 += vertex_weight[i];
+      } else {
+        balance_part_1 += vertex_weight[i];
+      }
+    }
+
+    float balance_ratio
+        = std::abs(balance_part_0 / (balance_part_0 + balance_part_1) - 0.5);
+    if (balance_ratio < balance_constraint) {
+      logger_->report("Partitioning successful with balance ratio of %.2f%%",
+                      balance_ratio * 100);
+      return true;
+    }
+
+    logger_->report("Partitioning failed with balance ratio of %.2f%%",
+                    balance_ratio * 100);
+    return false;
+  };
+
+  if (checkBalance() == false) {
```

Comment:
This appears to just be a workaround to the underlying TP bug.  I think @AcKoucher has identified the root cause as being bad initial partitioning.